### ISSUE TITLE
[SPMD] SPMD supports Gradient Accumulation in data parallel

### ISSUE THREAD

**Issue by baoleai** [openxla/xla#6750]
When using SPMD for data parallelism and applying gradient accumulation, the current version of XLA can introduce additional all-reduce operations. Is there a plan to support gradient accumulation under SPMD, similar to what is discussed in this GitHub issue: https://github.com/alpa-projects/alpa/issues/72? @tensorflower-gardener

**Comment by cheshire** [comment]
@jurahul @bchetioui WDYT?

**Comment by bchetioui** [comment]
If I understand correctly, I think this is already supposed to be supported (`WhileLoopAllReduceCodeMotion` should specifically take care of this). Can you confirm that this is on GPU @baoleai?

If so, are you able to provide a reproducer for your problem? There may be a bug somewhere.

**Comment by baoleai** [comment]
Thank you for your feedback @bchetioui. I understand that the `WhileLoopAllReduceCodeMotion` approach can handle the `allreduce + add` pattern in a while loop, but it is not applicable in the case as mentioned above in the torch_xla issue. In that case, the `allreduce + add` pattern is present in multiple sub-modules rather than in a while loop.

**Comment by bchetioui** [comment]
Thanks for getting back to me @baoleai. From what I can understand, the issue mentioned on the alpa repository is only dealing with while loops.

I'm not sure if I understand exactly what you mean by "multiple sub-modules". Is it perhaps related to the conditional control flow op, and by "sub-module" you mean different "HloComputation"s? That could make sense, cross-computation-boundary optimizations are not something we do afaik, and not something that is easy to do either.

If you could provide a HLO showing exactly what you want to fix, I'd be happy to try and give some pointers around how we could approach it. Otherwise, if you're already working on a fix, feel free to ask me for review once you're done :)
